pkg/types/workload.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package types
+
+import (
+	"fmt"
+	"github.com/cilium/cilium/api/v1/models"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"regexp"
+	"strings"
+)
+
+var (
+	cronJobNameRegexp = regexp.MustCompile(`(.+)-\d{8,10}$`)
+)
+
+// Workload Kubernetes pod’s workload info (workloads are: Deployment, Statefulset, Daemonset, ReplicationController, CronJob, Job, DeploymentConfig (OpenShift), etc).
+type Workload struct {
+	Name      string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
+	Kind      string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+}
+
+// Equal compare two workload is equal
+func (w *Workload) Equal(o *Workload) bool {
+	if w == o {
+		return true
+	}
+	if (w == nil && o != nil) || (w != nil && o == nil) {
+		return false
+	}
+	return w.Name == o.Name && w.Namespace == o.Namespace && w.Kind == o.Kind
+}
+
+// GetWorkloadFromModel  convert models.Workload to this workload struct
+func GetWorkloadFromModel(workload *models.Workload) *Workload {
+	if workload == nil {
+		return nil
+	}
+	return &Workload{
+		Name:      workload.Name,
+		Namespace: workload.Namespace,
+		Kind:      workload.Kind,
+	}
+}
+
+// GetWorkloadDataFromPod GetWorkloadMetaFromPod and cronJobNameRegexp are copied from
+// https://github.com/istio/istio/blob/1aca7a67afd7b3e1d24fafb2fbfbeaf1e41534c0/pkg/kube/util.go
+//
+// Modifications:
+// This code has been modified and move to this file from k8s/utils/workload.go
+// main changes are store the workload data in workload struct instead of slim_metav1.ObjectMeta and slim_metav1.TypeMeta
+// Below is the original comment in k8s/utils/workload.go:
+// GetDeployMetaFromPod has been renamed to GetWorkloadMetaFromPod and has
+// been updated to use the cilium slim API types.
+// We do not store the APIVersion of the owning workload in the TypeMeta
+// either, because it isn't needed for our purposes, and our slim types do not
+// have this field.
+// We fallback to the pod's ownerReference if we cannot find a more suitable
+// workload based on heuristics, whereas the original code defaulted to the
+// pod's name. This may be the case when using ReplicaSets without a Deployment.
+func GetWorkloadDataFromPod(pod *slim_corev1.Pod) (*Workload, error) {
```

Comment:
This mostly seems to be a copy of `utils.GetWorkloadMetaFromPod`. Can you reuse that function, or adapt it so the logic isn't duplicated between these two functions?

Reply:
Since original function `utils.GetWorkloadMetaFromPod` is only used in one place now. so I adapt that place  to use the new function and delete the function  `utils.GetWorkloadMetaFromPod`.